### ISSUE TITLE
linter: false positive for typescript(use-unknown-in-catch-callback-variable)

### ISSUE THREAD

**Issue by victorz** [oxc-project/oxc#22591]
### What version of Oxlint are you using?

1.66.0

### What command did you run?

`oxlint file.js`

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```ts
import { defineConfig } from "oxlint";

// oxlint-disable-next-line import/no-default-export
export default defineConfig({
  ignorePatterns: [
    "coverage/**",
    "dist/**",
    "docs/**",
    "scripts/**",
    "submodules/**",
    "sln/**",
    "test-results/**",
    "node_modules/**",
  ],
  plugins: [
    "eslint",
    "import",
    "jest",
    "jsdoc",
    "jsx-a11y",
    "node",
    "oxc",
    "promise",
    "react",
    "react-perf",
    "typescript",
    "unicorn",
    "vitest",
  ],
  categories: {
    correctness: "error",
    suspicious: "warn",
    pedantic: "warn",
    perf: "warn",
    style: "warn",
    restriction: "warn",
    nursery: "off",
  },
  options: {
    typeAware: true,
    maxWarnings: 25,
  },
  rules: {
    // deny
    "typescript/restrict-template-expressions": ["deny", { allowNumber: true }],

    // warn
    "eslint/max-params": ["warn", { max: 6 }],
    "eslint/max-statements": ["warn", { max: 30 }],
    "eslint/prefer-destructuring": ["warn", { array: false }],
    "unicorn/catch-error-name": ["warn", { ignore: ["cause"] }],
    "eslint/no-console": [
      "warn",
      { allow: ["info", "warn", "error", "assert"] },
    ],

    // allow
    ...Object.fromEntries(
      [
        "eslint/array-callback-return",
        "eslint/capitalized-comments",
        "eslint/default-case",
        "eslint/default-param-last",
        "eslint/func-style",
        "eslint/id-length",
        "eslint/max-lines",
        "eslint/max-lines-per-function",
        "eslint/max-statements",
        "eslint/no-bitwise",
        "eslint/no-empty-function",
        "eslint/no-implicit-coercion",
        "eslint/no-inline-comments",
        "eslint/no-magic-numbers",
        "eslint/no-negated-condition",
        "eslint/no-nested-ternary",
        "eslint/no-plusplus",
        "eslint/no-shadow",
        "eslint/no-ternary",
        "eslint/no-undefined",
        "eslint/no-underscore-dangle",
        "eslint/no-void",
        "eslint/no-warning-comments",
        "eslint/require-await",
        "eslint/sort-imports",
        "eslint/sort-keys",
        "import/consistent-type-specifier-style",
        "import/exports-last",
        "import/first",
        "import/group-exports",
        "import/max-dependencies",
        "import/no-cycle",
        "import/no-default-export",
        "import/no-named-as-default",
        "import/no-named-export",
        "import/no-namespace",
        "import/no-relative-parent-imports",
        "import/no-unassigned-import",
        "import/prefer-default-export", // disable
        "import/unambiguous",
        "jsdoc/require-param",
        "jsdoc/require-param-description",
        "jsdoc/require-param-name",
        "jsdoc/require-param-type",
        "jsdoc/require-returns",
        "jsdoc/require-returns-type",
        "oxc/branches-sharing-code",
        "oxc/no-async-await",
        "oxc/no-optional-chaining",
        "oxc/no-rest-spread-properties",
        "promise/always-return",
        "promise/catch-or-return",
        "promise/prefer-await-to-callbacks",
        "promise/prefer-await-to-then",
        "react-perf/jsx-no-jsx-as-prop",
        "react-perf/jsx-no-new-array-as-prop",
        "react-perf/jsx-no-new-function-as-prop",
        "typescript/consistent-return",
        "typescript/consistent-type-definitions",
        "typescript/dot-notation",
        "typescript/explicit-function-return-type",
        "typescript/explicit-module-boundary-types",
        "typescript/no-confusing-void-expression",
        "typescript/no-deprecated",
        "typescript/no-misused-promises",
        "typescript/no-unsafe-argument",
        "typescript/no-unsafe-assignment",
        "typescript/no-unsafe-call",
        "typescript/no-unsafe-member-access",
        "typescript/no-unsafe-return",
        "typescript/no-unsafe-type-assertion",
        "typescript/prefer-nullish-coalescing",
        "typescript/prefer-readonly-parameter-types",
        "typescript/promise-function-async",
        "typescript/require-await",
        "typescript/restrict-plus-operands",
        "typescript/strict-boolean-expressions",
        "typescript/strict-void-return",
        "unicorn/consistent-function-scoping",
        "unicorn/explicit-length-check",
        "unicorn/filename-case",
        "unicorn/new-for-builtins",
        "unicorn/no-array-callback-reference",
        "unicorn/no-array-for-each",
        "unicorn/no-array-reduce",
        "unicorn/no-array-sort",
        "unicorn/no-immediate-mutation",
        "unicorn/no-nested-ternary",
        "unicorn/no-null",
        "unicorn/no-object-as-default-parameter",
        "unicorn/no-useless-undefined",
        "unicorn/no-zero-fractions",
        "unicorn/numeric-separators-style",
        "unicorn/prefer-at",
        "unicorn/prefer-code-point",
        "unicorn/prefer-math-trunc",
        "unicorn/prefer-native-coercion-functions",
        "unicorn/prefer-string-replace-all",
        "unicorn/prefer-string-slice",
        "unicorn/switch-case-braces",
        // might be a good idea to look over these performance rules:
        "oxc/no-map-spread",
        "oxc/no-accumulating-spread",
      ].map((rule) => [rule, "allow"])
    ),

    // plugins react
    "react/jsx-max-depth": ["warn", { max: 10 }],
    "react/jsx-filename-extension": ["deny", { extensions: ["jsx", "tsx"] }],
    ...Object.fromEntries(
      [
        "jsx-a11y/prefer-tag-over-role",
        "react/display-name",
        "react/forbid-component-props",
        "react-in-jsx-scope",
        "react/jsx-curly-brace-presence",
        "react/jsx-key",
        "react/jsx-no-constructed-context-values",
        "react/jsx-props-no-spreading",
        "react/no-multi-comp",
        "react/only-export-components",
        // probably want to enable this:
        "react-perf/jsx-no-new-object-as-prop",
      ].map((rule) => [rule, "allow"])
    ),

    // plugins vitest, jest

    // We will prefer vitest/prefer-strict-boolean-matchers over these two
    // rules, but these are still allowed if they make sense, you just have
    // to disable the occurences with line comments.
    // Leaving these allowed for now, but they should be enabled by being
    // commented out.
    "vitest/prefer-to-be-falsy": "allow", // enable
    "vitest/prefer-to-be-truthy": "allow", // enable

    "jest/max-expects": ["warn", { max: 30 }],
    "vitest/max-expects": ["warn", { max: 30 }],
    ...Object.fromEntries(
      [
        "jest/consistent-test-it",
        "jest/no-conditional-in-test",
        "jest/no-untyped-mock-factory",
        "jest/padding-around-test-blocks",
        "jest/prefer-called-with",
        "jest/prefer-ending-with-an-expect",
        "jest/prefer-expect-assertions",
        "jest/prefer-importing-jest-globals",
        "jest/prefer-lowercase-title",
        "jest/prefer-mock-return-shorthand",
        "jest/prefer-spy-on",
        "jest/prefer-strict-equal",
        "jest/prefer-to-have-length",
        "jest/require-hook",
        "jest/require-top-level-describe",
        "jest/require-to-throw-message",
        "vitest/consistent-test-it",
        "vitest/no-conditional-in-test",
        "vitest/prefer-called-once", // enable
        "vitest/prefer-called-times",
        "vitest/prefer-describe-function-title",
        "vitest/prefer-expect-assertions",
        "vitest/prefer-importing-vitest-globals",
        "vitest/prefer-import-in-mock",
        "vitest/prefer-strict-boolean-matchers", // enable
        "vitest/require-mock-type-parameters",
        "vitest/require-test-timeout",
        "vitest/require-top-level-describe",
        "vitest/warn-todo", // keep -- fix some day!
      ].map((rule) => [rule, "allow"])
    ),
  },
});
```

### What happened?

I am getting a false positive of `typescript/use-unknown-in-catch-callback-variable` with this minimal reproducible example:

```js
  Promise.resolve()
    .then(() => {})
    .catch((error) => {
      console.log(error);
    });
```

The rule is triggered on the `error` parameter of the `.catch()` clause.

This is a JavaScript file, so I feel like the rule shouldn't be checked for at all -- we can't add type annotations in a JavaScript file. But maybe I've misconfigured oxlint in some way? If so, please let me know how I've misunderstood this thing! 😁️

**Comment by camc314** [comment]
Hi, thanks for reporting this issue.

Closing this out as not planned.

To resolve this, please add the type annotation (jsdoc) to the `error` parameter (detailed below)

```
Promise.resolve()
  .then(() => {})
  .catch((/** @type unknown*/ error) => {
    console.log(error);
  });

```

Alternatively, if you do not want to enable this type aware rule on a JS file add an override:
```
{
  "overrides": [
    {
      "files": ["*.js"],
      "rules": { "typescript/use-unknown-in-catch-callback-variable": "off" }
    }
}
```


Please let me know if this does not resolve the issue - thanks!

**Comment by victorz** [comment]
Thanks for that suggestion, @camc314! It does resolve this particular case. 🙏️

We are currently in a mixed JS + TS situation. Does this mean that in order to disable all the rules from the `typescript` plugin for JS files, we'd have to disable them each, one by one? Or could we perhaps in an override for `*.js` files just specify all the global plugins we use minus the `typescript` plugin? Would that work?

I tried something to that effect but I wasn't very successful. Should we perhaps be doing it the other way around? Enable all plugins globally _except_ for `typescript`, and then enable it in an override for `*.ts` files?

**Comment by camc314** [comment]
> Or could we perhaps in an override for *.js files just specify all the global plugins we use minus the typescript plugin? Would that work?

This works. see below for a demo:

https://github.com/camc314/oxlint-repros/tree/main/issue-22591

Please give it a try and let me know if it's working as expected.

Thanks!

**Comment by camc314** [comment]
> We are currently in a mixed JS + TS situation. 

To add another note here: there's not much harm in enabliong the typescript rules for `.js` files - there shouldn't be any false positives (unless TS can't infer the type), and enabling the rules should help catching more lint errors.

**Comment by victorz** [comment]
Brilliant, thank you very much! For reference for anyone encountering this, we removed the `typescript` plugin from the global plugins, then added this override:

```ts
  overrides: [
    {
      files: ["*.ts", "*.tsx"],
      plugins: ["typescript"],
      rules: {
        "typescript/typescript-specific-rule": "allow/warn/deny",
        ...
      },
    },
  ],
```

This means we can probably do the same for test files in the same way, adding an override and enabling test rules only for those files, etc. Very handy. 🙏️

Thanks again for your help, @camc314! ⭐️
